### ISSUE TITLE
Doubts about the inconsistency between the implementation in metacloak.py and the pseudo code Algorithm 1 in the paper

### ISSUE THREAD

**Issue by huzimun** [liuyixin-louis/MetaCloak#6]
Dear author, when I read the core code in metacloak.py, I found that the actual code and the implementation of the Inner-loop in pseudo-code are not consistent:

1. After the inner loop of the pseudo-code ends, the sum of losses of multiple agent models is calculated as a gradient (pseudo code 14-16), while the actual code adds a loop in the inner loop (actual code line583) and calculates the model gradient one-by-one;

2. When dealing with the lower level problem, the pseudo-code fine-tunes the model with unroll steps=1 (pseudo code line6-line8), but the actual code did not use unroll steps and fine-tunes the model with advance_steps=2 (actual code line588-line596).

I would like to ask you the following two questions:

1. Why can multi-model parallel simultaneous perturbation in pseudo-code be replaced by multi-model serial sequential perturbation in actual implementation? Is there any theoretical basis or reference?
2. Why must a new loop "for j in range (args. interval//args. advance_steps)" be added? What are the specific meanings of interval and advance_steps in the newly added loop, and how are their values determined?

Thank you for reading my issue in your busy schedule, and I look forward to your reply. 

Actual code in metacloak.py:
![Image](https://github.com/user-attachments/assets/12b4f95e-7cdd-4a77-ae3f-7951a14dc3b2)
Pseudocode in MetaCloak paper:
![Image](https://github.com/user-attachments/assets/e46c18a0-a923-476c-b6f5-7aad996d774e)

**Comment by liuyixin-louis** [comment]
Thanks for your questions. The current codebase only provides the sequential perturbation implementation. For the implementation of multi-model parallel simultaneous perturbation, you can refer to `https://github.com/VinAIResearch/Anti-DreamBooth/blob/main/attacks/ensemble_aspl.py` for the adaptation. For the interval and advance_steps, the interval is the length of each step-staggered surrogate in the paper, and advance_steps is the unrolling step K in the paper. Please refer to the context for details

> To achieve this, we propose to learn perturbation over a
pool of steps-staggered surrogates. Specifically, we train
M surrogate models {θj }M
j=1 as the initial point for the j-th
surrogate model. Given maximum training steps Nmax, the
j-th surrogate model θj is trained with Lrob
db for ⌊jNmax/M ⌋
steps on clean data from the pre-trained weight θ0. With this
pool, we then conduct C crafting outer loop; for each loop,
we sample a batch of surrogates and train them separately
on the poison. Upon the completion of inner-loop training,
we average the generation losses L∗
gen over ensembles and
update the protected images Xp with gradient ascent. Following MAML [12 ], a first-order gradient approximation is
used for efficient perturbation crafting

**Comment by huzimun** [comment]
Your reply has resolved my doubts, thank you!

**Comment by liuyixin-louis** [comment]
Closing as resolved. Thanks for the discussion!
